Extension/src/LanguageServer/settings.ts

```diff
@@ -452,6 +461,224 @@ export class CppSettings extends Settings {
+        }
+        vscode.workspace.applyEdit(edits).then(() => vscode.window.showTextDocument(document));
+    }
+
+    public async generateEditorConfig(): Promise<void> {
+        let document: vscode.TextDocument;
+        if (this.resource) {
+            // If a folder is open and '.editorconfig' exists at the root, use that.
+            const uri: vscode.Uri = vscode.Uri.joinPath(this.resource, ".editorconfig");
+            const edits: vscode.WorkspaceEdit = new vscode.WorkspaceEdit();
+            edits.createFile(uri, { ignoreIfExists: true, overwrite: false });
+            try {
+                await vscode.workspace.applyEdit(edits);
+                document = await vscode.workspace.openTextDocument(uri);
+            } catch (e) {
+                document = await vscode.workspace.openTextDocument();
+            }
+        } else {
+            document = await vscode.workspace.openTextDocument();
+        }
+        this.populateEditorConfig(document);
+    }
+
+    // If formattingEngine is set to "Default", searches for .editorconfig with vcFormat
+    // entries, or a .clang-format file, to determine which settings to use.
+    // This is intentionally not async to avoid races due to multiple entrancy.
+    public useVcFormat(document: vscode.TextDocument): boolean {
+        if (this.formattingEngine !== "Default") {
+            return this.formattingEngine === "vcFormat";
+        }
+        if (this.clangFormatStyle !== "file") {
+            // If a clang-format style other than file is specified, don't try to switch to vcFormat.
+            return false;
+        }
+        const cachedValue: boolean | undefined = cachedEditorConfigLookups.get(document.uri.fsPath);
+        if (cachedValue !== undefined) {
+            return cachedValue;
+        }
+        let foundEditorConfigWithVcFormatSettings: boolean = false;
+        const findConfigFile: (parentPath: string) => boolean = (parentPath: string) => {
+            const editorConfigPath: string = path.join(parentPath, ".editorconfig");
+            if (fs.existsSync(editorConfigPath)) {
+                const editorConfigSettings: any = getEditorConfigSettings(document.uri.fsPath);
+                const keys: string[] = Object.keys(editorConfigSettings);
+                for (let i: number = 0; i < keys.length; ++i) {
+                    if (keys[i].startsWith("cpp_")) {
+                        foundEditorConfigWithVcFormatSettings = true;
+                        const didEditorConfigNotice: PersistentState<boolean> = new PersistentState<boolean>("Cpp.didEditorConfigNotice", false);
+                        if (!didEditorConfigNotice.Value) {
+                            vscode.window.showInformationMessage(localize("editorconfig.default.behavior",
+                                "vcFormat was selected because an '.editorconfig' file was found containing vcFormat entries and 'Cpp.formatting' is set to 'Default'."));
```

Comment:
I think this message should only appear when both .editorconfig and .clang-format are detected in the same workspace, so we shouldn't pop it up here.

Message suggestion: "Code formatting will use settings from .editorconfig instead of .clang-format because the configuration is closer to this file. For more information, see the documentation for the 'C_Cpp.formatting' setting."

Reply:
(and we need to update `package.nls.json` with the more detailed information promised) 😄

Reply:
Note that the scan continues up the directory structure beyond the workspace folder (which is by design).  Once we find one, we don't currently continue scanning for the other.

Lack of both files would still default to `clangFormat`, so the behavior for the user would change whenever a `.editorconfig` is selected (regardless of whether we found a `.clang-format` as well).  So, I think it would make more sense to display a message any time `.editorconfig` is selected - any time (er, the first time) the behavior may have been changed out from under them.  If you agree, would you like to suggest new text?  :)

Reply:
"Code formatting will use settings from .editorconfig instead of .clang-format. For more information, see the documentation for the 'C_Cpp.formatting.Default' setting."

Documentation for the "Default" enum value should be something like: "By default, clang-format will be used to format the code. However, the Visual C++ formatting engine will be used if an .editorconfig file is found nearer to the code being formatted and clang_format_style is the default value: 'file'."